### ISSUE TITLE
The HMAC signing methods Sign method does not work with strings

### ISSUE THREAD

**Issue by bcokert** [dgrijalva/jwt-go#210]
When following the main example from the README, it says to use an HMAC signing method, and to then call SignedString on a created token.
```go
//https://godoc.org/github.com/dgrijalva/jwt-go#example-New--Hmac

// Create a new token object, specifying signing method and the claims
// you would like it to contain.
token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
    "foo": "bar",
    "nbf": time.Date(2015, 10, 10, 12, 0, 0, 0, time.UTC).Unix(),
})

// Sign and get the complete encoded token as a string using the secret
tokenString, err := token.SignedString(hmacSampleSecret)

fmt.Println(tokenString, err)
```
The natural thing to assume is that the key is a string. And since both SignedString (of token) and Sign (of HMAC) both accept an interface{}, the compiler doesn't warn about this.

However, using a string causes the following useless error during runtime:
```
key is invalid
```

And upon looking at the code that eventually throws this we see the problem:
```go
if keyBytes, ok := key.([]byte); ok {
```

This assertion will fail, as the input is of type string. However, since the parameter is an interface{}, one might think that the type assertion would be able to find a []byte inside it (took me a few minutes to realize why it didn't work).

After changing my input to a []byte, I then started seeing more useful errors in my tests:
```
key is of invalid type
```

Which implies that there is type checking of input keys in other areas of the code.

I'm not sure which way this should be resolved, as it could go one of 3 ways, and thus didn't create a PR. Here's what I think are the options:
1. Let the Sign method check for strings, and cast them to []byte. This special case is reasonable since []byte and string are always convertible, and it's a reasonable assumption that a key could be a string
2. Update the error message to be more useful, specifying that it's the type of the key that is the problem. In both cases that this error occurs in the repo the reason is the same (failed assertion to []byte), so it would be a quick fix
3. Add validation in a similar vein to other methods (token.Parse for example)

If one of these stands out as correct, let me know and I'll create a PR.

**Comment by dgrijalva** [comment]
This is one of the most common places people get stuck.  I believe the appropriate resolution is documentation. Users shouldn't have to guess the correct key type.  The challenge is that each signing method uses different key types.  I've found, though several iterations on this, that being as specific as possible with types ends up being better than being clever/convenient.

tl;dr; I think the right solution is to fix documentation.

**Comment by jayhuang75** [comment]
@dgrijalva I still try to figure out, so is that mean I have to base64 encoding my secret key? and convert to the byte?
`token, err := jwt.Parse(parts[1], func(token *jwt.Token) (interface{}, error) {
			// Don't forget to validate the alg is what you expect:
			if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
				return nil, fmt.Errorf("Unexpected signing me thod: %v", token.Header["alg"])
			}
			fmt.Println(token)

			my_secret_key := []byte("my_sercet_key")
			base64_encoded_key := base64.StdEncoding.EncodeToString(my_secret_key)
			// hmacSampleSecret is a []byte containing your secret, e.g. []byte("my_secret_key")
			return []byte(base64_encoded_key), nil
		})`
Thank you to guide me through this!

**Comment by dustinevan** [comment]
I'm stuck.

**Comment by dgrijalva** [comment]
TODO: update the documentation for 3.2.0 release

**Comment by dgrijalva** [comment]
Added documentation in two places and returning a more useful error when you supply the wrong value.
